### DISCUSSION TITLE
has anyone got madflight running in a fixed wing single engine plane?

### DISCUSSION THREAD

**Discussion by slenderman-poon** [qqqlab/madflight#26] [General]
wondering if anyone is currently running madflight in a plane, and if they would be willing to share the code?

**Comment by qqqlab** [comment]
This was actually one of my goals when I started this project, but up to now did not get to it. Also, I'm not the best plane pilot, and I will probably crash a lot during development, which doesn't help...

One thing that probably needs attention is the AHRS system. It works well for copters, probably because copters have a steady hover state. For planes I doubt that sensor fusion works when the plane is loitering (flying constant circles). I expect the current AHRS to under-report the bank angle in this case. I noticed in the INAV source code that they use some sort of compensation for this, but didn't look in detail what they are doing exactly.

**Reply by slenderman-poon** [reply]
I gather then it's not a simple case of twiddling a few of the variables in one of the example files then 😞
I'm in the process of building the hardware, waiting for a couple pieces to arrive from China. happy to be a crash test bunny just in case there is some code you could throw my way to test 😊

**Reply by qqqlab** [reply]
Cool, I like that! I got behind the keyboard and created a plane example from scratch. It would be great if you can test it. Probably just dry run only, or try to fly it in PASSTHRU mode. And if you're really brave you could try ANGLE mode. (See comments in the plane example for details.)

Keep an eye on updates, I might do some additional work on it this weekend.

**Reply by slenderman-poon** [reply]
damn, you da boss! to just sit down and smash out 600 lines like that is incredible 😱🙇‍♂️
i'd better get building 😊 this is going to need a new plane 😁

**Reply by slenderman-poon** [reply]
and more updates 😱 looking forward to testing the fly by wire 🥰
dumb question... the pin numbers in the code (I'm making this with and ESP32 module, not a dev kit) are they the IO numbers or the physical pin number on the dev board?

**Comment by slenderman-poon** [comment]
decided to take a step back and prototype instead of reinventing wheels... does this look correct for a madflight_board_default_ESP32s.h file?

//This pin layout is optimized for ESP32s 30 pin board, use "ESP32 Dev Module" as board in Arduino IDE

#define HW_BOARD_NAME "DEFAULT ESP32 BOARD - ESP32s DevKit 30 pin" 
#define HW_MCU "ESP32" //ESP32 - Most pins can be assigned freely

//NOTE: DON'T USE SAME PIN TWICE. All pins here get configured, even if they are not used. Set pin to -1 to disable.

//LED:
#ifndef HW_PIN_LED
  #define HW_PIN_LED                2 //Note: ESP32s has an on-board LED
#endif
#ifndef HW_LED_ON
  #define HW_LED_ON                 1 //0:low is on, 1:high is on
#endif

//IMU SPI:
#ifndef HW_PIN_SPI_MOSI
  #define HW_PIN_SPI_MOSI          23 //   defaults: VSPI 23, HSPI 13
#endif
#ifndef HW_PIN_SPI_MISO
  #define HW_PIN_SPI_MISO          19 //VP defaults: VSPI 19, HSPI 12
#endif
#ifndef HW_PIN_SPI_SCLK
  #define HW_PIN_SPI_SCLK          18 //   defaults: VSPI 18, HSPI 14
#endif
#ifndef HW_PIN_IMU_CS
  #define HW_PIN_IMU_CS            5  //   defaults: VSPI  5, HSPI 15
#endif
#ifndef HW_PIN_IMU_EXTI
  #define HW_PIN_IMU_EXTI          39 // VN silkscreen label
#endif

//BARO/MAG I2C:
#ifndef HW_PIN_I2C_SDA
  #define HW_PIN_I2C_SDA           21 //default: Wire 21
#endif
#ifndef HW_PIN_I2C_SCL
  #define HW_PIN_I2C_SCL           22 //default: Wire 22
#endif

//Outputs:
#ifndef HW_OUT_COUNT
  #define HW_OUT_COUNT             8
#endif
#ifndef HW_PIN_OUT_LIST
  #define HW_PIN_OUT_LIST {33,25,26,27,14,12,13,15} //for ESP32 it is recommended to use only pins 2,4,12-19,21-23,25-27,32-33 for motors/servos
#endif

//Serial Debug on tx0 (pin 1), rx0 (pin 3) connected to serial->USB converter

//RC Receiver:
#ifndef HW_PIN_RCIN_RX
  #define HW_PIN_RCIN_RX           35 //also used as PPM input
#endif
#ifndef HW_PIN_RCIN_TX
  #define HW_PIN_RCIN_TX           32
#endif

//GPS:
#ifndef HW_PIN_GPS_RX
  #define HW_PIN_GPS_RX            17 
#endif
#ifndef HW_PIN_GPS_TX
  #define HW_PIN_GPS_TX            16
#endif

//Battery ADC
#ifndef HW_PIN_BAT_V
  #define HW_PIN_BAT_V             34
#endif
#ifndef HW_PIN_BAT_I
  #define HW_PIN_BAT_I             -1
#endif

//BlackBox SPI:
#ifndef HW_PIN_SPI2_MISO
  #define HW_PIN_SPI2_MISO         -1
#endif
#ifndef HW_PIN_SPI2_MOSI
  #define HW_PIN_SPI2_MOSI         -1
#endif
#ifndef HW_PIN_SPI2_SCLK
  #define HW_PIN_SPI2_SCLK         -1
#endif
#ifndef HW_PIN_BB_CS
  #define HW_PIN_BB_CS             -1
#endif

const int HW_PIN_OUT[] = HW_PIN_OUT_LIST;
![ESP32s](https://github.com/user-attachments/assets/8d39932a-d150-4235-996b-0cac34994723)

**Reply by qqqlab** [reply]
That is one way of doing it, creating new header file. You could also uncomment the pin section in plane.ino and set your gpio numbers there.

This is my test setup on an old bix3

![IMG_2013](https://github.com/user-attachments/assets/8ad0bca8-2702-414b-b4bc-187eea0b419e)
![IMG_2012](https://github.com/user-attachments/assets/7c425f8e-4d41-47af-9fa0-a25c1549bf68)

**Reply by qqqlab** [reply]
Note the MPU module directly soldered to the ESP32S3 board, it makes stuff easy to mount in the plane: I glued it upside down below the wing. The ESP32 is powered from the ESC 5V BEC output via a diode.
